### PR TITLE
Confluence and Causal Invariance

### LINKED ISSUE #458: Short note on Confluence and Causal Invariance
### The problem
There is a common misconception that confluence and causal invariance as defined in Jonathan Gorard's papers are equivalent. They are not, in fact, and we need a research note explaining that.

### Possible solution
This should be a very short research note just containing two counterexamples. One for a rule that is confluent but not causal invariant. And the other one for a rule that is causal invariant but not confluent.

### PR DESCRIPTION
**Pull request by maxitg**
## Changes
* Closes #458.
* Adds a note with counterexamples showing that causal invariance and confluence are not equivalent, and do not imply one another.

## Comments
* There is some misinformation currently going on, this note is an attempt to address it.

## Rendered version
* Rendered version: [Confluence and Causal Invariance](https://github.com/maxitg/SetReplace/blob/confluenceAndCausalInvariance/Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md).

---
This change is [Reviewable](https://reviewable.io/reviews/maxitg/setreplace/479)

### CODE REVIEW

File: Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md
Comment (daneelsan) [reply Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md:141]: How about: "To do that, we need to implement..."

File: Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md
Comment (maxitg) [reply Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md:39]: According to the definition of confluence from above, the states only need to converge if the evolutions have started from a single state in the past. So, I kept this definition consistent. (I'm going to put the statement about it in the definition itself though.)

File: Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md
Comment (maxitg) [reply Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md:96]: I'm going to improve the wording slightly.

**Comment by maxitg** [comment]
@daneelsan, please use Reviewable for reviews (link is always at the bottom of the PR message). It makes it much easier to track comments.

File: Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md
Comment (maxitg) [reply Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md]: Done.

File: Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md
Comment (maxitg) [reply Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md]: Done.

File: Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md
Comment (maxitg) [reply Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md]: Done.

File: Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md
Comment (maxitg) [reply Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md]: Changed the wording.

File: Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md
Comment (maxitg) [reply Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md]: Done.

**Review by daneelsan — COMMENTED** [review]
> Reviewed 4 of 5 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/maxitg/setreplace/479)* status: all files reviewed, 5 unresolved discussions (waiting on @aokellermann, @daneelsan, and @taliesinb)

**Review by daneelsan — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/maxitg/setreplace/479)* status: all files reviewed, 3 unresolved discussions (waiting on @aokellermann, @daneelsan, and @taliesinb)

___
*[Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md, line 39 at r1](https://reviewable.io/reviews/maxitg/setreplace/479#-MJiUJBxS2LH8QdyObB9-r1-39:-MJj1Ykp60HaTwo4JLr4:bb6ywmx) ([raw file](https://github.com/maxitg/setreplace/blob/8adb804562bfcf6c2dbfa2faf80c5529b96cfa41/Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md#L39)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

LGTM
___
*[Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md, line 96 at r1](https://reviewable.io/reviews/maxitg/setreplace/479#-MJiUJBxS2LH8QdyObB9-r1-96:-MJj1Su38hSmeTyqxMeY:bb6ywmx) ([raw file](https://github.com/maxitg/setreplace/blob/8adb804562bfcf6c2dbfa2faf80c5529b96cfa41/Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md#L96)):*
<details><summary><i>Previously, maxitg (Max Piskunov) wrote…</i></summary><blockquote>

Changed the wording.
</blockquote></details>

LGTM

**Comment by maxitg** [comment]
I am going to wait for a review from @taliesinb and a possible review from Wolfram internal DQA before merging.

File: Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md
Comment (taliesinb) [inline Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md:20]: I was a little confused here because I expected you to quote the glossary's definition of confluence, but it doesn't have one. Maybe it's wroth adding a sentence to say "The glossary does not define confluence, but we can assume it has the standard meaning and the one given in Gorard's paper".

**Review by taliesinb — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/maxitg/setreplace/479)* status: all files reviewed, 7 unresolved discussions (waiting on @aokellermann, @daneelsan, and @maxitg)

___
*[Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md, line 25 at r2](https://reviewable.io/reviews/maxitg/setreplace/479#-MJqXEMpBAbQ_0b18pGQ:-MJqXEMpBAbQ_0b18pGR:b9kzos3) ([raw file](https://github.com/maxitg/setreplace/blob/02e7116ee3ea4f04f5cf60d70fc20f117aaff205/Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md#L25)):*
> ```Markdown
> > A state `a` is deemed confluent if, for all pairs of states `b`, `c` that can be reached from `a`, there exists a
> > state `d` that can be reached from both `b` and `c`.
> > If every state in the system is confluent, the system itself is confluent.
> ```

A small diagram here to illustrate confluence vs causal invariance would be nice, but that's totally optional and shouldn't block a PR.
___
*[Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md, line 65 at r2](https://reviewable.io/reviews/maxitg/setreplace/479#-MJqYQQd7Q6ehUFt4UnN:-MJqYQQd7Q6ehUFt4UnO:b4erdui) ([raw file](https://github.com/maxitg/setreplace/blob/02e7116ee3ea4f04f5cf60d70fc20f117aaff205/Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md#L65)):*
> ```Markdown
> ```
> 
> The system is confluent, as any partial evolution, if continued, will always terminate at the final state `{{1}}`:
> ```

Just FYI I am probably going to suggest in https://github.com/maxitg/SetReplace/issues/427 that we standardize on the terms _{partial ,}history_ instead of _{partial ,}evolution_.
___
*[Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md, line 135 at r2](https://reviewable.io/reviews/maxitg/setreplace/479#-MJqZQQDEOvnWKUCxQ0P:-MJqZQQDEOvnWKUCxQ0Q:b-w2njli) ([raw file](https://github.com/maxitg/setreplace/blob/02e7116ee3ea4f04f5cf60d70fc20f117aaff205/Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md#L135)):*
> ```Markdown
> 
> ## Future Research
> ```

An optional addition:

I would add another idea to the mix here, which is that we can define a *colored* form of causal invariance that requires isomorphism of *colored causal graphs: causal graphs whose edges are colored by the index of the rule that applied to them. I have not thought about this at all, but it could be this can salvage the identity `confluence ⇔ causal invariance`. 

If that can be obviously falsified, then it is still perhaps worth speculating on *what* properties (or kinds of properties) are needed to make the identity false or true, and how we might further think about that issue.

**Review by maxitg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/maxitg/setreplace/479)* status: 1 of 9 files reviewed, 7 unresolved discussions (waiting on @aokellermann, @daneelsan, and @taliesinb)

___
*[Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md, line 20 at r2](https://reviewable.io/reviews/maxitg/setreplace/479#-MJiUJBxS2LH8QdyObB9-r2-20:-MJrZE7o8vAdGwEks5Do:bl8bokc) ([raw file](https://github.com/maxitg/setreplace/blob/02e7116ee3ea4f04f5cf60d70fc20f117aaff205/Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md#L20)):*
<details><summary><i>Previously, taliesinb (Taliesin Beynon) wrote…</i></summary><blockquote>

I was a little confused here because I expected you to quote the glossary's definition of confluence, but it doesn't have one. Maybe it's wroth adding a sentence to say "The glossary does not define confluence, but we can assume it has the standard meaning and the one given in Gorard's paper". 
</blockquote></details>

It's not from Jonathan's paper, it's from the wiki, I added clarification as suggested.
___
*[Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md, line 25 at r2](https://reviewable.io/reviews/maxitg/setreplace/479#-MJqXEMpBAbQ_0b18pGQ:-MJs6lMcAl5gpOjDpNPS:bxcdxfe) ([raw file](https://github.com/maxitg/setreplace/blob/02e7116ee3ea4f04f5cf60d70fc20f117aaff205/Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md#L25)):*
<details><summary><i>Previously, taliesinb (Taliesin Beynon) wrote…</i></summary><blockquote>

A small diagram here to illustrate confluence vs causal invariance would be nice, but that's totally optional and shouldn't block a PR.
</blockquote></details>

Done (see below).
___
*[Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md, line 65 at r2](https://reviewable.io/reviews/maxitg/setreplace/479#-MJqYQQd7Q6ehUFt4UnN:-MJr_8cL8UyqUeUj6kXq:b19dylq) ([raw file](https://github.com/maxitg/setreplace/blob/02e7116ee3ea4f04f5cf60d70fc20f117aaff205/Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md#L65)):*
<details><summary><i>Previously, taliesinb (Taliesin Beynon) wrote…</i></summary><blockquote>

Just FYI I am probably going to suggest in https://github.com/maxitg/SetReplace/issues/427 that we standardize on the terms _{partial ,}history_ instead of _{partial ,}evolution_.
</blockquote></details>

Not that we are using *evolution* in the code, in particular, `WolframModelEvolutionObject`, so this will affect reverse compatibility. We will have to rename it to `HistoryObject` if we are to do this.
___
*[Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md, line 135 at r2](https://reviewable.io/reviews/maxitg/setreplace/479#-MJqZQQDEOvnWKUCxQ0P:-MJr_ioo8BoE6Sqh7Vd3:bqiycrd) ([raw file](https://github.com/maxitg/setreplace/blob/02e7116ee3ea4f04f5cf60d70fc20f117aaff205/Research/ConfluenceAndCausalInvariance/ConfluenceAndCausalInvariance.md#L135)):*
<details><summary><i>Previously, taliesinb (Taliesin Beynon) wrote…</i></summary><blockquote>

An optional addition:

I would add another idea to the mix here, which is that we can define a *colored* form of causal invariance that requires isomorphism of *colored causal graphs: causal graphs whose edges are colored by the index of the rule that applied to them. I have not thought about this at all, but it could be this can salvage the identity `confluence ⇔ causal invariance`. 

If that can be obviously falsified, then it is still perhaps worth speculating on *what* properties (or kinds of properties) are needed to make the identity false or true, and how we might further think about that issue.
</blockquote></details>

Updated the examples to use a single rule only to avoid the first objection. Mentioned the second suggestion.

**Review by taliesinb — APPROVED** [review]
> Reviewed 8 of 8 files at r3.
> *[Reviewable](https://reviewable.io/reviews/maxitg/setreplace/479)* status: all files reviewed, 5 unresolved discussions (waiting on @aokellermann, @daneelsan, and @taliesinb)

**Review by maxitg — COMMENTED** [review]
Waiting until next week for Wolfram DQA.

> *[Reviewable](https://reviewable.io/reviews/maxitg/setreplace/479#-:-MK5idROCeYt8ofHgSMK:b-r0ai76)* status: all files reviewed, 5 unresolved discussions (waiting on @aokellermann, @daneelsan, and @taliesinb)

**Comment by maxitg** [comment]
Wolfram DQA decided to review using notebooks, so I'm merging this for now. Any possible issues they find will be addressed in other PRs.

**Pull request merged**
